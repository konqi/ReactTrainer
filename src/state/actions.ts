export enum ActionType {
  ADD_TRAINEE
}
