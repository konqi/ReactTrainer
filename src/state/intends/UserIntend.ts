import {Trainee} from '../../types/Trainee'
import {FSA} from '../../types/FSA'
import {Session} from '../../types/Session'

export enum UserIntend {
  ADD_TRAINEE = '[INTEND] add trainee',
  DELETE_TRAINEE = '[INTEND] delete trainee',
  ADD_SESSION = '[INTEND] add session',
  SHOW_TRAINEES = '[INTEND] show trainees',
  SHOW_TRAINEE_DETAILS = '[INTEND] show trainee details',
}

export type AddTraineeIntendPayload = Omit<Trainee, 'id'>
export type AddTraineeIntendFSA = FSA<
  UserIntend.ADD_TRAINEE,
  AddTraineeIntendPayload
>
export const createAddTraineeIntend = (
  payload: AddTraineeIntendPayload
): AddTraineeIntendFSA => ({
  type: UserIntend.ADD_TRAINEE,
  payload,
})

export const createShowTraineesIntend = () => ({
  type: UserIntend.SHOW_TRAINEES,
})

export type AddSessionIntendPayload = Omit<Session, 'id' | 'traineeRef'>
export interface AddSessionIntendMeta {
  traineeId: string
}
export type AddTrainingIntendFSA = FSA<
  UserIntend.ADD_SESSION,
  AddSessionIntendPayload,
  AddSessionIntendMeta
>
export const createAddTrainingIntend = (
  training: AddSessionIntendPayload,
  traineeId: string
): AddTrainingIntendFSA => ({
  type: UserIntend.ADD_SESSION,
  payload: training,
  meta: {traineeId},
})

export type ShowTraineeDetailsIntendFSA = FSA<
  UserIntend.SHOW_TRAINEE_DETAILS,
  string
>
export const createShowTraineeDetailsIntend = (
  traineeId: string
): ShowTraineeDetailsIntendFSA => ({
  type: UserIntend.SHOW_TRAINEE_DETAILS,
  payload: traineeId,
})
