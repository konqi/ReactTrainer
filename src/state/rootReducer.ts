import {ApplicationState} from '.'
import {FSA} from '../types/FSA'
import {reduceTrainees} from './trainees'

export const rootReducer = (
  state: ApplicationState,
  action: FSA<any, any>
): ApplicationState => ({
  trainees: reduceTrainees(state.trainees, action),
})
