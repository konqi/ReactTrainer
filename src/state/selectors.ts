import {ApplicationState} from '.'

export const selectTraineeById = (state: ApplicationState, traineeId: string) =>
  state.trainees[traineeId]
