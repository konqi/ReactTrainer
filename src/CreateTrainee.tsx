import { Button, TextField } from "@material-ui/core";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import * as React from "react";
import { ApplicationContext } from "./App";
import { ActionType } from "./state/actions";
import { useState, useContext } from "react";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    container: {
      display: "grid",
      gridTemplateRows: "auto auto auto",
      justifyItems: "center"
    },
    textField: {
      marginLeft: theme.spacing(1),
      marginRight: theme.spacing(1),
      width: 200
    },
    dense: {
      marginTop: 19
    },
    menu: {
      width: 200
    }
  })
);

export const CreateTrainee = () => {
  const classes = useStyles();
  const [name, setName] = useState("");
  const [price, setPrice] = useState<number>(1500);
  const { dispatch } = useContext(ApplicationContext);

  return (
    <form className={classes.container} noValidate autoComplete="off">
      <TextField
        id="nameTextField"
        label="Name"
        className={classes.textField}
        value={name}
        onChange={({ target: { value } }) => {
          setName(value);
        }}
        margin="normal"
      />
      <TextField
        id="priceTextField"
        label="Preis"
        className={classes.textField}
        value={String(price)}
        onChange={({ target: { value } }) => {
          const num = Number(value);
          isNaN(num) || (isFinite(num) && setPrice(num));
        }}
        margin="normal"
      />
      <Button
        onClick={() => {
          dispatch!({
            type: ActionType.ADD_TRAINEE,
            payload: {
              name,
              price
            }
          });
        }}
        variant="contained"
        color="primary"
      >
        Anlegen
      </Button>
    </form>
  );
};
