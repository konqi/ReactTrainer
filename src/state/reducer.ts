import { ActionType } from "./actions";

export interface Action {
  type: ActionType;
  payload: Trainee;
}

export interface Trainee {
  name: string;
  price: number;
}

export interface ApplicationState {
  trainees: Trainee[];
}

export const initialApplicationState: ApplicationState = {
  trainees: []
};

export const reducer = (
  state: ApplicationState,
  action: Action
): ApplicationState => {
  switch (action.type) {
    case ActionType.ADD_TRAINEE:
      const newTrainees = [...state.trainees, action.payload];
      return { ...state, trainees: newTrainees };
    default:
      throw new Error();
  }
};
