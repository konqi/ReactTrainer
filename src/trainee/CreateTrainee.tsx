import {Button, TextField} from '@material-ui/core'
import {createStyles, makeStyles, Theme} from '@material-ui/core/styles'
import * as React from 'react'
import {useState} from 'react'
import {useDispatch} from 'react-redux'
import {createAddTraineeIntend} from '../state/intends/UserIntend'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    container: {
      display: 'grid',
      gridTemplateRows: 'auto auto auto',
      justifyItems: 'center',
    },
    textField: {
      // marginLeft: theme.spacing(1),
      // marginRight: theme.spacing(1),
    },
  })
)

export const CreateTrainee: React.FC = () => {
  const classes = useStyles()
  const [name, setName] = useState('')
  const [price, setPrice] = useState<number>(15)
  const dispatch = useDispatch()

  return (
    <form className={classes.container} noValidate autoComplete="off">
      <TextField
        id="nameTextField"
        label="Name"
        className={classes.textField}
        value={name}
        onChange={({target: {value}}) => {
          setName(value)
        }}
        fullWidth
        margin="normal"
      />
      <TextField
        id="priceTextField"
        type="number"
        label="Preis"
        className={classes.textField}
        value={String(price)}
        onChange={({target: {value}}) => {
          if (typeof value !== 'string') {
            return
          }
          if ((value as string).length < 1) {
            return
          }
          const num = Number(value)
          isNaN(num) || (isFinite(num) && setPrice(num))
        }}
        fullWidth
        margin="normal"
      />
      <Button
        onClick={() =>
          dispatch(
            createAddTraineeIntend({
              name,
              price,
            })
          )
        }
        variant="contained"
        color="primary"
      >
        Anlegen
      </Button>
    </form>
  )
}
