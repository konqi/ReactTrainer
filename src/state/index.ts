import {TraineeState, TraineeFSAs} from './trainees/types'
import {initialTraineeState} from './trainees'
import {UiState, UiFSAs} from './ui/types'
import {TrainingFSAs} from './trainings/types'
import {initialUiState} from './ui/uiReducer'
import {createStore, applyMiddleware} from 'redux'
import {rootReducer} from './rootReducer'
import createSagaMiddleware from 'redux-saga'
import {rootSaga} from './sagas/rootSaga'

export interface ApplicationState {
  trainees: TraineeState
  ui: UiState
}

export const initialApplicationState: ApplicationState = {
  trainees: initialTraineeState,
  ui: initialUiState,
}

export type Action = TraineeFSAs | TrainingFSAs | UiFSAs

const sagaMiddleware = createSagaMiddleware()

export const store = createStore(
  rootReducer,
  initialApplicationState,
  applyMiddleware(sagaMiddleware)
)

sagaMiddleware.run(rootSaga)
