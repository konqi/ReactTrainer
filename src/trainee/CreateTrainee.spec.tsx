import {fireEvent, render} from '@testing-library/react'
import * as React from 'react'
import * as ReactRedux from 'react-redux'
import {Provider} from 'react-redux'
import {store} from '../state'
import {createAddTraineeIntend} from '../state/intends/UserIntend'
import {CreateTrainee} from './CreateTrainee'

it('snapshot test', () => {
  const {baseElement, unmount} = render(
    <Provider store={store}>
      <CreateTrainee />
    </Provider>
  )

  expect(baseElement).toMatchSnapshot()

  unmount()
})

describe('integration tests', () => {
  let dispatch = jest.fn()
  let useDispatch: any
  beforeEach(() => {
    dispatch.mockReset()
    useDispatch = jest
      .spyOn(ReactRedux, 'useDispatch')
      .mockImplementation(() => dispatch)
  })
  afterEach(() => {
    useDispatch.mockRestore()
  })

  test('loads items eventually', async () => {
    const {getByText, getByLabelText, unmount} = render(
      <Provider store={store}>
        <CreateTrainee />
      </Provider>
    )
    const expectedName = 'Dr. Testilon'
    fireEvent.change(getByLabelText('Name'), {
      target: {value: expectedName},
    })

    const expectedPrice = 2000
    fireEvent.change(getByLabelText('Preis'), {
      target: {value: expectedPrice},
    })

    // Click button
    fireEvent.click(getByText('Anlegen'))

    expect(dispatch).toHaveBeenCalledWith(
      createAddTraineeIntend({name: expectedName, price: expectedPrice})
    )

    unmount()
  })
})

describe('unit tests', () => {
  test.each`
    input     | expected
    ${'4711'} | ${'4711'}
    ${'ABC'}  | ${''}
    ${'!'}    | ${''}
    ${'0'}    | ${''}
    ${'0815'} | ${'815'}
  `(
    'when input is "$input" output should be "$expected"',
    ({input, expected}) => {
      const {getByLabelText, unmount} = render(
        <Provider store={store}>
          <CreateTrainee />
        </Provider>
      )
      const priceInput = getByLabelText('Preis') as HTMLInputElement

      fireEvent.change(priceInput, {
        target: {value: input},
      })
      expect(priceInput.value).toBe(expected)

      unmount()
    }
  )
})
