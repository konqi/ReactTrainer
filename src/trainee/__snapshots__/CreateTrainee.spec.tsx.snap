// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<body>
  <div>
    <form
      autocomplete="off"
      class="makeStyles-container-1"
      novalidate=""
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-textField-2 MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="nameTextField"
        >
          Name
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="nameTextField"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-textField-2 MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink="true"
          for="priceTextField"
        >
          Preis
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="priceTextField"
            type="number"
            value="15"
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Anlegen
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</body>
`;
